File: python3/2901.py
class Solution:
    def validHam(self, i, j):
        if len(i) != len(j):
            return False
        f = False
        for k in range(0, len(i)):
            if i[k] != j[k]:
                if f:
                    return False
                f = True
        return True

    def find(self, a, g, i, n, l):
        if i == n:
            return
        self.find(a, g, i + 1, n, l)
        p = []
        for k in range(i + 1, n):
            if g[i] != g[k] and self.validHam(a[i], a[k]) and len(p) < len(l[k]):
                p = l[k]
        l[i] = [a[i]] + p

    def getWordsInLongestSubsequence(self, w: List[str], g: List[int]) -> List[str]:
        d = [[] for _ in range(len(w))]
        self.find(w, g, 0, len(w), d)
        t = d[0]
        for i in d:
            if len(t) < len(i):
                t = i
        return t

File: java/2901.java
class Solution {
    private boolean validHam(char i[], char j[]) {
        if (i.length != j.length)
            return false;
        int n = i.length;
        boolean f = false;
        for (int k = 0; k < n; k++)
            if (i[k] != j[k]) {
                if (f)
                    return false;
                f = true;
            }
        return true;
    }

    private void find(char a[][], int g[], int i, int n, List<String> l[]) {
        if (i == n)
            return;
        List<String> t = new ArrayList<>();
        l[i] = t;
        t.add(new String(a[i]));
        find(a, g, i + 1, n, l);
        List<String> p = null;
        for (int k = i + 1; k < n; k++)
            if (g[i] != g[k] && validHam(a[i], a[k]) && (p == null || p.size() < l[k].size()))
                p = l[k];
        if (p != null)
            t.addAll(p);
    }

    public List<String> getWordsInLongestSubsequence(String[] w, int[] g) {
        int n = w.length;
        char a[][] = new char[n][];
        for (int i = 0; i < n; i++)
            a[i] = w[i].toCharArray();
        List<String> d[] = new List[n];
        find(a, g, 0, n, d);
        List<String> t = d[0];
        for (List<String> k : d)
            if (t.size() < k.size())
                t = k;
        return t;
    }
}
